node/actors/bft/src/tests/unit_tests.rs

```diff
@@ -0,0 +1,599 @@
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_eq!(res, Ok(()));
+    let leader_prepare = util.recv_signed().await.unwrap();
+    let res = util.dispatch_leader_prepare(leader_prepare).await;
+
+    assert_eq!(res, Ok(()));
+}
+
+#[tokio::test]
+async fn leader_prepare_invalid_leader() {
+    let mut util = Util::make_with(2).await;
+
+    let view = ViewNumber(2);
+    util.set_replica_view(view);
+    util.set_leader_view(view);
+
+    assert_eq!(util.view_leader(view), util.key_at(0).public());
+
+    let replica_prepare_one = util.make_replica_prepare(None::<fn(&mut ReplicaPrepare)>);
+    let res = util.dispatch_replica_prepare(replica_prepare_one.clone());
+    assert_eq!(
+        res,
+        Err(ReplicaPrepareError::NumReceivedBelowThreshold {
+            num_messages: 1,
+            threshold: 2,
+        })
+    );
+
+    let replica_prepare_two = util.key_at(1).sign_msg(replica_prepare_one.msg);
+    let res = util.dispatch_replica_prepare(replica_prepare_two);
+    assert_eq!(res, Ok(()));
+
+    let mut leader_prepare = util.recv_leader_prepare().await.unwrap();
+    leader_prepare.view = leader_prepare.view.next();
+    assert_ne!(
+        util.view_leader(leader_prepare.view),
+        util.key_at(0).public()
+    );
+    let leader_prepare = util
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare));
+    let res = util.dispatch_leader_prepare(leader_prepare).await;
+
+    assert_eq!(
+        res,
+        Err(LeaderPrepareError::InvalidLeader {
+            correct_leader: util.key_at(1).public(),
+            received_leader: util.key_at(0).public(),
+        })
+    );
+}
+
+#[tokio::test]
+async fn leader_prepare_invalid_sig() {
+    let mut util = Util::make().await;
+
+    let mut leader_prepare = util.make_leader_prepare(None::<fn(&mut LeaderPrepare)>);
+    leader_prepare.sig = util.rng().gen();
+    let res = util.dispatch_leader_prepare(leader_prepare).await;
+
+    assert_eq!(
+        res,
+        Err(LeaderPrepareError::InvalidSignature(
+            SignatureVerificationFailure
+        ))
+    );
+}
+
+#[tokio::test]
+async fn leader_prepare_invalid_prepare_qc_different_views() {
+    let mut util = Util::make().await;
+
+    let replica_prepare = util.make_replica_prepare(None::<fn(&mut ReplicaPrepare)>);
+    let res = util.dispatch_replica_prepare(replica_prepare.clone());
+    assert_eq!(res, Ok(()));
+
+    let mut leader_prepare = util.recv_leader_prepare().await.unwrap();
+    leader_prepare.view = leader_prepare.view.next();
+    let leader_prepare = util
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare));
+    let res = util.dispatch_leader_prepare(leader_prepare).await;
+
+    assert_eq!(
+        res,
+        Err(LeaderPrepareError::InvalidPrepareQC(anyhow!(
+            "PrepareQC contains messages for different views!"
+        )))
```

Comment:
Makes little sense to compare errors exactly. I *think* in this case matching an error to `InvalidPrepareQC` would be enough. If it's not enough, the error itself should be refactored, so that this variant wraps a more strongly typed error instead of `anyhow::Error`.

Reply:
+1

Reply:
But if you are going to move some tests to the `roles` crate anyway, the best is to move from `anyhow` to `thiserror` like you had to do in the `consensus` crate.

Reply:
Yes, it was just a temp solution. 
Both solutions will be applied.

Reply:
Yes, it was just a temp solution.
Both suggestions will be applied.